testing/monkey/monkey_dev.c

```diff
@@ -0,0 +1,325 @@
+      free(dev);
+    }
+
+  return NULL;
+}
+
+/****************************************************************************
+ * Name: monkey_dev_delete
+ ****************************************************************************/
+
+void monkey_dev_delete(FAR struct monkey_dev_s *dev)
+{
+  MONKEY_ASSERT_NULL(dev);
+
+  if (dev->fd > 0)
+    {
+      /* Reset input state */
+
+      struct monkey_dev_state_s state;
+      memset(&state, 0, sizeof(state));
+      state.type = dev->type;
+      monkey_dev_set_state(dev, &state);
+
+      MONKEY_LOG_NOTICE("close fd: %d", dev->fd);
+      close(dev->fd);
+    }
+
+  free(dev);
+}
+
+/****************************************************************************
+ * Name: monkey_dev_set_state
+ ****************************************************************************/
+
+void monkey_dev_set_state(FAR struct monkey_dev_s *dev,
+                          FAR const struct monkey_dev_state_s *state)
+{
+  MONKEY_ASSERT_NULL(dev);
+
+  switch (MONKEY_GET_DEV_TYPE(dev->type))
+    {
+      case MONKEY_DEV_TYPE_TOUCH:
+        utouch_write(dev->fd,
+                    state->data.touch.x,
+                    state->data.touch.y,
+                    state->data.touch.is_pressed);
+        break;
+
+      case MONKEY_DEV_TYPE_BUTTON:
+        ubutton_write(dev->fd, state->data.button.value);
+        break;
+
+      default:
+        break;
+    }
+}
+
+/****************************************************************************
+ * Name: monkey_dev_get_state
+ ****************************************************************************/
+
+bool monkey_dev_get_state(FAR struct monkey_dev_s *dev,
+                          FAR struct monkey_dev_state_s *state)
+{
+  bool retval;
+  MONKEY_ASSERT_NULL(dev);
+
+  retval = false;
+
+  state->type = dev->type;
+
+  switch (dev->type)
+    {
+      case MONKEY_DEV_TYPE_TOUCH:
+        retval = touch_read(dev->fd,
+                            &state->data.touch.x,
+                            &state->data.touch.y,
+                            &state->data.touch.is_pressed);
+        break;
+
+      case MONKEY_DEV_TYPE_BUTTON:
+        retval = button_read(dev->fd, &state->data.button.value);
+        break;
+
+      default:
+        break;
+    }
+
+  return retval;
+}
+
+/****************************************************************************
+ * Name: monkey_dev_get_state
+ ****************************************************************************/
+
+enum monkey_dev_type_e monkey_dev_get_type(
+                       FAR struct monkey_dev_s *dev)
+{
+  MONKEY_ASSERT_NULL(dev);
+  return dev->type;
+}
+
+/****************************************************************************
+ * Name: monkey_dev_get_available
+ ****************************************************************************/
+
+int monkey_dev_get_available(FAR struct monkey_dev_s *devs[], int dev_num)
+{
+  int i;
+  int available = 0;
+  struct pollfd fds[MONKEY_DEV_MAX_NUM];
```

Comment:
Do we need `memset(&fds, 0, sizeof(fds));`?

Reply:
done